### DISCUSSION TITLE
Timemap: proposal to represent score time as a fraction in each JSON object

### DISCUSSION THREAD

**Discussion by yucongj** [rism-digital/verovio#3663] [Ideas]
Currently in [Timemap](https://book.verovio.org/toolkit-reference/output-formats.html#timemap), the score time is somewhat represented in a field called `qstamp`, which is a number representing "the time in quarter notes from the start of the music to the start of the current event entry". For example (the last example in [Timemap](https://book.verovio.org/toolkit-reference/output-formats.html#timemap)), the notation below would result in `qstamp`s: `0`, `1.333333`, `2.666667`, and `4` (offset).

![image](https://github.com/rism-digital/verovio/assets/60488433/85d33091-23df-45fb-ba41-fc6c12695cd0)

However, representing score times as float numbers loses information when e.g., rounding $\frac{4}{3}$ to be `1.333333`. This is problematic in applications where precise score times (as fractions) are needed, as converting from floats back to fractions could be ambiguious, and may result in incorrect fractions like $\frac{171}{128}$.

Since **every** score time is essentially a fraction (however complicated) that can be unambiguiously generated from the MEI file, why not preserve this information in Timemap? :-)  I propose adding two fields to Timemap: (i) `cfraction`: a fraction representing the cumulative music time of a JSON object, **with a quarter note worth $\frac{1}{4}$**; and  (ii) `mfraction`: a fraction representing this object's local position within the measure. The field names and other details can be determined in the future, but here's an example for demonstrating the idea:

The table below shows the values of `qstamp`, `cfraction`, and `mfraction` at the beginning of [Beethoven_Op31_No3](https://github.com/trompamusic-encodings/Beethoven_Op31_No3_HenleUrtext/blob/master/Beethoven_Op31_No3_1-HenleUrtext.mei), and then the two triplets in Measure 8. For readability, the first column specifies the measure number (though this also could be included as a new field).

![Screenshot 2024-05-02 at 11 49 15 PM](https://github.com/rism-digital/verovio/assets/60488433/470ebeae-dd6e-4d7b-ac52-10585f2da330)

| measure | qstamp | cfraction | mfraction |
| --- | --- | --- | --- |
| 1 | 0 | 0/1 | 0/1 |
| 1 | 0.75 | 3/16 | 3/16 |
| 1 | 1 | 1/4 | 1/4 |
| 1 | 2 | 1/2 | 1/2 |
| 2 | 3 | 3/4 | 0/1 |
| 2 | 3.75 | 15/16 | 3/16 |
| 2 | 4 | 1/1 | 1/4 |
| 2 | 5 | 5/4 | 1/2 |
| 3 | 6 | 3/2 | 0/1 |
| 3 | 7 | 7/4 | 1/4 |
| 3 | 8 | 2/1 | 1/2 |
| ... | ... | ... | ... |
| 8 | 22 | 11/2 | 1/4 |
| 8 | 22.33333333333333 | 67/12 | 1/3 |
| 8 | 22.66666666666667 | 17/3 | 5/12 |
| 8 | 23 | 23/4 | 1/2 |
| 8 | 23.33333333333333 | 35/6 | 7/12 |
| 8 | 23.66666666666667 | 71/12 | 2/3 |
| 9 | 24 | 6/1 | 0/1 |
| ... | ... | ... | ... |

Such fractions can be directly calculated from the MEI file, maining using each note's `dur` attribute (e.g., `dur=8` represents an eighth note), and when present, using the `tuplet` tag as well (which in this case would interpret `dur=8` as the value `1/12` instead of `1/8`).

I believe including these fractions in the Timemap output is a generic feature that can benefit a variety of applications, especially the ones that involve time stamps in performance recordings. This would certainly benefit my project (which involves audio-to-score alignment!) with improved robustness and easier debugging.

**What do you all think?**

**Comment by lpugin** [comment]
This requires changing the alignment positions into fractions internally. This is a big change, but I can look into it.

**Reply by yucongj** [reply]
If you're willing to look into it, that would be great! I'm happy to discuss details offline if needed.

**Comment by craigsapp** [comment]
It is a good idea in general, although there would be two main concerns:  

(1) it increases the verbosity(size) of the timemap, 

(2) it increases the computational complexity of dealing with rational numbers.   The main complication with rational numbers is the need to do prime factorization to create reduced fractions, such as "3/6" to "1/2".

The proposed names could be improved: "cfraction" is long and does not show its relationship to qstamp.  `qfrac` would be better, for example.   The time units between `cfraction` (using whole notes as the unit) and qstamp (using quarter notes as the unit) are different, and this will cause confusion.  The representations "0/1" and "1/1", etc. are pedantic and I would suggest omitting the denominator when it is 1.

--------------------------

That being said, I use rational numbers for rhythm processing in humlib, and I don't find any noticeable problems with computing efficiency.  

Here is a verovio issue that was probably related to using floats instead of rational numbers:

https://github.com/rism-digital/verovio/issues/256
https://github.com/rism-digital/verovio/issues/263

There have not been any significant problems other than this I think, so not completely necessary, but using rational numbers for rhythm alignment would be more elegant since there will be no round-off error problem to compensate for.  

Rational numbers can be converted to floating-point numbers as needed after rhythm alignment, such as when dealing with `text@tstamp` which are floating-point values.   The floating-point numbers generated from rational numbers will have no problems with round-off errors, since the floating-point versions will be exactly the same


 `HumNum` is the rational number class used in humlib:

https://github.com/craigsapp/humlib/blob/master/include/HumNum.h
https://github.com/craigsapp/humlib/blob/master/src/HumNum.cpp


Other than implementing a rational number class (which I have already done, but see Rational class below), converting the alignment system to rational numbers should take about a day, with most of that involving testing rather than implementation.   The `HumNum` class has all of the math operators defined (at least those that will be needed), so most of the code would not change other than the variable type.


I had ChatGPT create a rational number class which seems to cover the essentials of what is needed (and which tells me about std::gcd existing in the numeric package, which I did not know about):

```cpp
#include <iostream>
#include <numeric>  // For std::gcd in C++17 and later

class Rational {
private:
    int numerator;
    int denominator;

    // Helper function to reduce the fraction
    void reduce() {
        int gcd = std::gcd(numerator, denominator);
        numerator /= gcd;
        denominator /= gcd;
        if (denominator < 0) {  // Make denominator positive
            numerator = -numerator;
            denominator = -denominator;
        }
    }

public:
    Rational(int num = 0, int denom = 1) : numerator(num), denominator(denom) {
        if (denom == 0) {
            throw std::invalid_argument("Denominator cannot be zero");
        }
        reduce();
    }

    // Addition
    Rational operator+(const Rational& rhs) const {
        return Rational(numerator * rhs.denominator + rhs.numerator * denominator,
                        denominator * rhs.denominator);
    }

    // Subtraction
    Rational operator-(const Rational& rhs) const {
        return Rational(numerator * rhs.denominator - rhs.numerator * denominator,
                        denominator * rhs.denominator);
    }

    // Multiplication
    Rational operator*(const Rational& rhs) const {
        return Rational(numerator * rhs.numerator, denominator * rhs.denominator);
    }

    // Division
    Rational operator/(const Rational& rhs) const {
        if (rhs.numerator == 0) {
            throw std::invalid_argument("Division by zero");
        }
        return Rational(numerator * rhs.denominator, denominator * rhs.numerator);
    }

    // Display function
    void display() const {
        std::cout << numerator << "/" << denominator << std::endl;
    }
};

int main() {
    Rational r1(3, 4);  // 3/4
    Rational r2(2, 5);  // 2/5

    Rational r3 = r1 + r2;
    r3.display();  // 23/20

    Rational r4 = r1 - r2;
    r4.display();  // 7/20

    Rational r5 = r1 * r2;
    r5.display();  // 6/20 -> 3/10

    Rational r6 = r1 / r2;
    r6.display();  // 15/8

    return 0;
}
```

**Reply by yucongj** [reply]
Thanks for your support for the idea! For concern 1), we could make the default timemap to not include such information, but only include it when users indicate so in optional arguments. For 2), we could also leave "3/6" be, instead of converting it to "1/2". Users could easily apply the GCD algorithm on their own to convert all fractions to the simplest form, as you demonstrated.

**Comment by craigsapp** [comment]
> . For 2), we could also leave "3/6" be, instead of converting it to "1/2". Users could easily apply the GCD algorithm on their own to convert all fractions to the simplest form, as you demonstrated.

I think the computational effort to apply GCD is worth it to reduce the fractions.   Fractions are useful for getting precise alignments for complicated tuplet rhythms in the calculation of music notation as well as for using in the timemap.  @lpugin is working on converting the double system for rhythm alignment, which will then make fractions available for timemap at no additional time cost.

>  1), we could make the default timemap to not include such information, but only include it when users indicate so in optional arguments. 

This is an interesting idea.  Currently there are three options to control the contents of the timemap (1) exclude cue notes, which is set with `--midi-no-cue` (2) include measure information and (3) include rests.   For options (2) and (3) I cannot remember how they are set (not through specific options, but embedded in a separate option?).

I propose adding an additional option such as `--timemap-fractions` to output the timemap using fraction values for qstamp.  Having both double and Fraction types at the same time is not very useful, since converting from Fraction to double is easy, so one possibility is to use the fraction option to switch from double values for qstamp to Fraction values.  Otherwise the `--timemap-fraction` option could output both, with the fraction parameter being `qfrac`.

It might be nice to have separate verovio options for measures and rests inclusions in the timemap (or make it easier to figure out how to set those parameters by mentioning them in the `verovio -h full` list of options.)  For example `--timemap-rests` and `--timemap-measures`.

**Reply by lpugin** [reply]
I added some documentation about the timemap options [here](https://book.verovio.org/toolkit-reference/toolkit-methods.html#rendertotimemap)

One issue is that we currently cannot change them when using the command line tool. One reason for using JSON options in such cases is not to multiply Verovio options. For this reason, I would prefer not to have a new Verovio option for each of them. However, we should quite likely have one that is command-line only and that allows to specify them. Something like `--timemap-options`. However, when thinking about this, I was never sure if it would be better to have the value being a JSON string, or a JSON file. Any thoughts? Maybe we can allow both and auto detect if the value passed starts with '{' ?

**Reply by craigsapp** [reply]
Yes, that sounds good.   Can quasi-JSON be used, or in other words without requiring quotes around the parameters keys like a Javascript Object?:

`--timemap-options "{measures: true, rests: true, fractions: true}"`

as opposed to 

`--timemap-options '{"measures": true, "rests": true, "fractions": true}'`

and/or allowing `1` in the place of `true`:

`--timemap-options "{measures: 1, rests: 1, fractions: 1}"`

In that case whitespace followed by `{` would be a good search for differentiate, loading from a file:

`--timemap-options my-options.json`

But the later could be done similar to engraving options:

`--timemap-options-file my-options.json`

similar to `--engraving-defaults-file engraving.json`

>  is not to multiply Verovio options

I think that is too late 😉

**Reply by lpugin** [reply]
It would need to be using quotes. But `true` should work. So it would be:
```
--timemap-options '{"measures": true, "rests": true, "fractions": true}'
```

**Comment by craigsapp** [comment]
How to interact with the `--midi-no-cue` for MIDI output?   Should that setting be shared between MIDI and timemap options (currently they are the same, which might be best to leave).   This could be a way of overriding the MIDI output option when creating Timemaps:

`--timemap-options "{measures: true, rests: true, fractions: true, nocue: true}"`

On the command-line this would not matter since only MIDI or only timemap will be created before the toolkit is exited, but for a persistent toolkit, the options might be more complicated to handle when creating both MIDI and timemaps.

**Comment by craigsapp** [comment]
> `--timemap-options '{"measures": true, "rests": true, "fractions": true}'`

That is OK.  In the documentation for command-line use, it would be helpful to explain that the outer quotes should be single and the inner quotes double (or vice versa).  Otherwise back slashes would be needed if they are both double quotes (and maybe also works with single quotes):

 `--timemap-options "{\"measures\": true, \"rests\": true, \"fractions\": true}"`

**Reply by lpugin** [reply]
I think users using the command-line are aware of this.

**Reply by craigsapp** [reply]
> I think users using the command-line are aware of this.

It is OK to omit, but fewer know than you may realize 🥲

**Comment by craigsapp** [comment]
A question is how to implement the Fraction option for timemap.  Do you have a preference @yucongj (or others)?

Currently there is one timemap parameters called `qstamp` (meaning cumulative quarter-note duration from the start of the score).

* The `qstamp` should remain double value as is currently being done for backwards compatibility and for real-time processing for highlighting notes in the score, as is the primary purpose of the timemap.

* How to add the factional qstamp value can be done in a few ways:

(1) Add a new parameter called `qfrac` which gives the Fraction version of `qstamp`.   This could be always added by default, although I would prefer it to be added only by an option for creating timemap such as (command-line version): `--timemap-options '{"fractions": true}'`

(2) Alternatively, the data type of `qstamp` could be changed from double to Fraction using this option:

 `--timemap-options '{"useFractions": true}'`

In this case there is only `qstamp`, but it will be expressed in Fractions rather than the default of doubles.

-----------

In general, users will probably only want doubles or Fractions, so the second one seems a good idea unless both are needed at the same time.   But this would complicate automated cases since you would not necessarily know the data type of `qstamp`.  But I cannot think of cases where this would be a significant problem, since the double can be calculated easily from the Fraction form.

For maximum flexibility it would be best to have an option that adds `qfrac` parameters for the Fraction value and always include `qstamp` for the double version.   Then the timemap option could be:

 `--timemap-options '{"includeFractions": true}'`

This would add the `qfrac` (or `qfraction`) parameter, and the `qstamp` would also be present with the double form of the cumulative quarter note duration.

There is also the request for the fractional duration from the start of the last barline.   This could be called `mfrac` or `mfraction` and would give the cumulative quarter-note duration since the last barline, and perhaps these can be included at the same time when `includeFractions` option is given.

**Reply by lpugin** [reply]
I am planning to:
* Keep double as default since this is probably going to remain the most widely used format and not to break compatibility
* Have the option to be `"useFractions": true` and keep it `qstamp` in the timemap.

Regarding the option to have the values reset for each measure, I am not sure about that one because it can be calculated. In fact, this is independent from the use of double or fraction, so we would possibly have another option to trigger that.

**Reply by lpugin** [reply]
One more point. Why not putting the fraction into an array?
```json
"qstamp": [
    1,
    2 
]
```
Since using fractions will quite likely be for some analysis calculation, this will avoid the need to do string parsing of `"1/2"`

@yucongj ? What do you think?

**Reply by craigsapp** [reply]
That sounds good to me.

**Reply by yucongj** [reply]
I think it's better to call the fractions `qfrac` rather than the same name as the double `qstamp`, because it could be confusing to call two different data representations the same name. 

> In general, users will probably only want doubles or Fractions, so the second one seems a good idea unless both are needed at the same time. But this would complicate automated cases since you would not necessarily know the data type of qstamp. But I cannot think of cases where this would be a significant problem, since the double can be calculated easily from the Fraction form.

Even though we cannot think of such problematic cases right now, it might still cause unexpected problems in the future, so why not nip it in the bud? More likely so, if we put the fraction into an array (which I have no problem with!).

I also think `qfrac` is just a more accurate name for the fractions. `qstamp` inherently sounds like a double, like in the word "timestamp" which is usually a double value (—but maybe it's just me?). **Therefore, I'd prefer using the option `"fractions": true` to include both the original `qstamp` and the new `qfrac`**.

If we convert `qfrac` to double (and multiplying it by 4.0), the result should match `qstamp` exactly; but if somehow they don't match at some (strange) places, it'd also be easier to debug when both of them are present.

It's a judgement call.

**Comment by craigsapp** [comment]
> If we convert qfrac to double (and multiplying it by 4.0), the result should match qstamp exactly; but if somehow they don't match at some (strange) places, it'd also be easier to debug when both of them are present.

For `qfrac`, it must be in units of quarter notes due to the `q` prefix.  If you want it in units of whole notes, then it should be something like `wfrac` for whole-note fractions.  But in general mixed units should be avoided between similar parameters, so it is better to have `qfrac` which would be expressed in quarter-note units.  The receiving application would divide by 4 to generate `wfrac` data.

For me, I do not have any particular preference for expressing `qstamp` as either a double or a Fraction as compared to keeping them separate with `qfrac` being the Fraction form of `qstemp`.  A particular usage case may indicate a particular preference.  In general the creation of the timemap will be done on demand, so the context will be understood by the purpose it is being created (you would create the Fraction version and then use the information immediately).  If timemaps are stored for later use, perhaps that will cause complications.  But it will not be too difficult: you can check the `qstamp` value to see if it is an array or a number to verify the format.

If there is a usage for having both such as wanting Fractions primarily, but then use it for realtime highlighting, the one usage would be to create the Fraction form, and then later calculate the double version on the fly, either dynamically from the Fraction value, or to preprocess the Fraction and create a separate entry with double.  In that case being able to use `qstamp` to add the double value may be useful.  But in such cases another parameter name could be used for the double form of `qtstamp` for the specific application.

Outside the context of verovio, I have done another variant for accessing both doubles and fractions, which would be something like this:

```json
"qstamp": [4.5, 1, 2]
```

The first number is the double value, and the second/third numbers are the Fraction part of the non-integer part of the double:

```
int(4.5) + 1/2  = 4 + 1/2 = 9/2.
```

This allows easier understanding of the Fraction value since `4 + 1/2` is easier to understand the meaning of compared to `9/2`.  In particular it would be easier to factor the values 1 and 2 in this case when reading the number compared to 9/2, which looks like it has something to do with triplets, when it usually will not.

**Reply by lpugin** [reply]
I am OK with naming the key `qfrac`. However, I am really not convinced we need both, especially if we give the fraction as an array of numbers.

Having the tempo as a number makes sense - it should have been like that in the first place. This is a backward change, it might break some applications but it will probably remain unnoticed in most JS applications. A note will be added to the changelog.

I updated the [PR](https://github.com/rism-digital/verovio/pull/3816) accordingly
